In [1]:
import pandas as pd
import numpy as np
import pyreadr
import math
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf

In [2]:
rdata_read = pyreadr.read_r("C:/Users/edfra/OneDrive/Documentos/GitHub/ECO224/Labs/data/wage2015_subsample_inference.Rdata")

# Extracting the data frame from rdata_read
df = rdata_read[ 'data' ]

df.shape

(5150, 20)

In [3]:
df

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we  \
rownames                                                                     
10         9.615385  2.263364  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
12        48.076923  3.872802  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
15        11.057692  2.403126  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
18        13.942308  2.634928  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
19        28.846154  3.361977  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
...             ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
32620     14.769231  2.692546  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
32624     23.076923  3.138833  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0   
32626     38.461538  3.649659  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   
32631     32.967033  3.495508  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32643     17.307692  2.851151  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   

           ne  exp1  exp2    exp3     exp4   occ occ2   ind ind2  
rownames                                                          
10        1.0   7.0  0.49   0.343   0.2401  3600   11  8370   18  
12        1.0  31.0  9.61  29.791  92.3521  3050   10  5070    9  
15        1.0  18.0  3.24   5.832  10.4976  6260   19   770    4  
18        1.0  25.0  6.25  15.625  39.0625   420    1  6990   12  
19        1.0  22.0  4.84  10.648  23.4256  2015    6  9470   22  
...       ...   ...   ...     ...      ...   ...  ...   ...  ...  
32620     0.0   9.0  0.81   0.729   0.6561  4700   16  4970    9  
32624     0.0  12.0  1.44   1.728   2.0736  4110   13  8680   20  
32626     0.0  11.0  1.21   1.331   1.4641  1550    4  3680    6  
32631     0.0  10.0  1.00   1.000   1.0000  2920    9  6570   11  
32643     0.0  14.0  1.96   2.744   3.8416  1610    5  7460   14  

[5150 rows x 20 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5150 entries, 10 to 32643
Data columns (total 20 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   wage    5150 non-null   float64 
 1   lwage   5150 non-null   float64 
 2   sex     5150 non-null   float64 
 3   shs     5150 non-null   float64 
 4   hsg     5150 non-null   float64 
 5   scl     5150 non-null   float64 
 6   clg     5150 non-null   float64 
 7   ad      5150 non-null   float64 
 8   mw      5150 non-null   float64 
 9   so      5150 non-null   float64 
 10  we      5150 non-null   float64 
 11  ne      5150 non-null   float64 
 12  exp1    5150 non-null   float64 
 13  exp2    5150 non-null   float64 
 14  exp3    5150 non-null   float64 
 15  exp4    5150 non-null   float64 
 16  occ     5150 non-null   category
 17  occ2    5150 non-null   category
 18  ind     5150 non-null   category
 19  ind2    5150 non-null   category
dtypes: category(4), float64(16)
memory usage: 736.3+ KB


In [5]:
df.describe()

wage        lwage          sex          shs          hsg  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean     23.410410     2.970787     0.444466     0.023301     0.243883   
std      21.003016     0.570385     0.496955     0.150872     0.429465   
min       3.021978     1.105912     0.000000     0.000000     0.000000   
25%      13.461538     2.599837     0.000000     0.000000     0.000000   
50%      19.230769     2.956512     0.000000     0.000000     0.000000   
75%      27.777778     3.324236     1.000000     0.000000     0.000000   
max     528.845673     6.270697     1.000000     1.000000     1.000000   

               scl          clg           ad           mw           so  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.278058     0.317670     0.137087     0.259612     0.296505   
std       0.448086     0.465616     0.343973     0.438464     0.456761   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     1.000000     0.000000     1.000000     1.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

                we           ne         exp1         exp2         exp3  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.216117     0.227767    13.760583     3.018925     8.235867   
std       0.411635     0.419432    10.609465     4.000904    14.488962   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     5.000000     0.250000     0.125000   
50%       0.000000     0.000000    10.000000     1.000000     1.000000   
75%       0.000000     0.000000    21.000000     4.410000     9.261000   
max       1.000000     1.000000    47.000000    22.090000   103.823000   

              exp4  
count  5150.000000  
mean     25.118038  
std      53.530225  
min       0.000000  
25%       0.062500  
50%       1.000000  
75%      19.448100  
max     487.968100

In [6]:
data_shs = df[df[ 'shs' ] == 1]
data_shs

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we  \
rownames                                                                      
500         8.547009  2.145581  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
540        10.439560  2.345602  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
691        14.903846  2.701619  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
843         9.615385  2.263364  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1775        9.442308  2.245200  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
...              ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
29976     528.845673  6.270697  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
30178      15.865385  2.764140  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
31119      13.461538  2.599837  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
32075      43.269231  3.767442  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   
32143      24.000962  3.178094  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   

           ne  exp1     exp2       exp3        exp4    occ occ2    ind ind2  
rownames                                                                     
500       1.0  37.0  13.6900  50.653000  187.416100   6800   19    490    2  
540       1.0  26.0   6.7600  17.576000   45.697600  1e+05   21  1e+05    5  
691       1.0  29.0   8.4100  24.389000   70.728100   4220   14   7070   13  
843       1.0  28.0   7.8400  21.952000   61.465600   4020   13   8680   20  
1775      1.0   8.0   0.6400   0.512000    0.409600   4230   14   7690   16  
...       ...   ...      ...        ...         ...    ...  ...    ...  ...  
29976     0.0  19.0   3.6100   6.859000   13.032100   4760   16   4690    9  
30178     0.0  18.0   3.2400   5.832000   10.497600  1e+05   21   2980    6  
31119     0.0  19.0   3.6100   6.859000   13.032100   4700   16   4970    9  
32075     0.0  25.5   6.5025  16.581375   42.282506   6400   19    770    4  
32143     0.0  15.0   2.2500   3.375000    5.062500   8140   21   8870   21  

[120 rows x 20 columns]

In [7]:
data_hsg = df[df[ 'hsg' ] == 1]
data_hsg

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we  \
rownames                                                                     
15        11.057692  2.403126  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
43        19.230769  2.956512  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
44        19.230769  2.956512  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
47        12.000000  2.484907  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
73        17.307692  2.851151  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
...             ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
32580     12.980769  2.563469  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32590     13.461538  2.599837  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32599     22.596154  3.117780  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32603     16.826923  2.822980  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32631     32.967033  3.495508  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   

           ne  exp1   exp2    exp3      exp4   occ occ2   ind ind2  
rownames                                                            
15        1.0  18.0   3.24   5.832   10.4976  6260   19   770    4  
43        1.0  42.0  17.64  74.088  311.1696  5120   17  7280   14  
44        1.0  37.0  13.69  50.653  187.4161  5240   17  5680    9  
47        1.0  31.0   9.61  29.791   92.3521  4040   13  8590   19  
73        1.0   7.0   0.49   0.343    0.2401  4020   13  8270   18  
...       ...   ...    ...     ...       ...   ...  ...   ...  ...  
32580     0.0  15.0   2.25   3.375    5.0625  2010    6  9370   22  
32590     0.0   8.0   0.64   0.512    0.4096  4720   16  8590   19  
32599     0.0  15.0   2.25   3.375    5.0625  9620   22  5390    9  
32603     0.0  11.0   1.21   1.331    1.4641  7150   20  8770   21  
32631     0.0  10.0   1.00   1.000    1.0000  2920    9  6570   11  

[1256 rows x 20 columns]

In [8]:
frames1 = [data_shs,data_hsg]

In [9]:
df1 = pd.concat(frames1)
df1

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we  \
rownames                                                                     
500        8.547009  2.145581  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
540       10.439560  2.345602  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
691       14.903846  2.701619  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
843        9.615385  2.263364  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1775       9.442308  2.245200  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
...             ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
32580     12.980769  2.563469  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32590     13.461538  2.599837  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32599     22.596154  3.117780  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32603     16.826923  2.822980  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32631     32.967033  3.495508  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   

           ne  exp1   exp2    exp3      exp4    occ occ2    ind ind2  
rownames                                                              
500       1.0  37.0  13.69  50.653  187.4161   6800   19    490    2  
540       1.0  26.0   6.76  17.576   45.6976  1e+05   21  1e+05    5  
691       1.0  29.0   8.41  24.389   70.7281   4220   14   7070   13  
843       1.0  28.0   7.84  21.952   61.4656   4020   13   8680   20  
1775      1.0   8.0   0.64   0.512    0.4096   4230   14   7690   16  
...       ...   ...    ...     ...       ...    ...  ...    ...  ...  
32580     0.0  15.0   2.25   3.375    5.0625   2010    6   9370   22  
32590     0.0   8.0   0.64   0.512    0.4096   4720   16   8590   19  
32599     0.0  15.0   2.25   3.375    5.0625   9620   22   5390    9  
32603     0.0  11.0   1.21   1.331    1.4641   7150   20   8770   21  
32631     0.0  10.0   1.00   1.000    1.0000   2920    9   6570   11  

[1376 rows x 20 columns]

In [10]:
data = df1[["wage","lwage","sex","ad","mw","so","we","ne","exp1","exp2","exp3","exp4","occ","occ2","ind","ind2",]]
data

wage     lwage  sex   ad   mw   so   we   ne  exp1   exp2  \
rownames                                                                   
500        8.547009  2.145581  0.0  0.0  0.0  0.0  0.0  1.0  37.0  13.69   
540       10.439560  2.345602  0.0  0.0  0.0  0.0  0.0  1.0  26.0   6.76   
691       14.903846  2.701619  0.0  0.0  0.0  0.0  0.0  1.0  29.0   8.41   
843        9.615385  2.263364  0.0  0.0  0.0  0.0  0.0  1.0  28.0   7.84   
1775       9.442308  2.245200  1.0  0.0  0.0  0.0  0.0  1.0   8.0   0.64   
...             ...       ...  ...  ...  ...  ...  ...  ...   ...    ...   
32580     12.980769  2.563469  0.0  0.0  0.0  0.0  1.0  0.0  15.0   2.25   
32590     13.461538  2.599837  1.0  0.0  0.0  0.0  1.0  0.0   8.0   0.64   
32599     22.596154  3.117780  0.0  0.0  0.0  0.0  1.0  0.0  15.0   2.25   
32603     16.826923  2.822980  0.0  0.0  0.0  0.0  1.0  0.0  11.0   1.21   
32631     32.967033  3.495508  0.0  0.0  0.0  0.0  1.0  0.0  10.0   1.00   

            exp3      exp4    occ occ2    ind ind2  
rownames                                            
500       50.653  187.4161   6800   19    490    2  
540       17.576   45.6976  1e+05   21  1e+05    5  
691       24.389   70.7281   4220   14   7070   13  
843       21.952   61.4656   4020   13   8680   20  
1775       0.512    0.4096   4230   14   7690   16  
...          ...       ...    ...  ...    ...  ...  
32580      3.375    5.0625   2010    6   9370   22  
32590      0.512    0.4096   4720   16   8590   19  
32599      3.375    5.0625   9620   22   5390    9  
32603      1.331    1.4641   7150   20   8770   21  
32631      1.000    1.0000   2920    9   6570   11  

[1376 rows x 16 columns]

In [11]:
education = np.ones((1376,1))
education

array([[1.],
       [1.],
       [1.],
       ...,
       [1.],
       [1.],
       [1.]])

In [12]:
data['education'] = education

<ipython-input-12-6c018b1083aa>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['education'] = education


In [13]:
data

wage     lwage  sex   ad   mw   so   we   ne  exp1   exp2  \
rownames                                                                   
500        8.547009  2.145581  0.0  0.0  0.0  0.0  0.0  1.0  37.0  13.69   
540       10.439560  2.345602  0.0  0.0  0.0  0.0  0.0  1.0  26.0   6.76   
691       14.903846  2.701619  0.0  0.0  0.0  0.0  0.0  1.0  29.0   8.41   
843        9.615385  2.263364  0.0  0.0  0.0  0.0  0.0  1.0  28.0   7.84   
1775       9.442308  2.245200  1.0  0.0  0.0  0.0  0.0  1.0   8.0   0.64   
...             ...       ...  ...  ...  ...  ...  ...  ...   ...    ...   
32580     12.980769  2.563469  0.0  0.0  0.0  0.0  1.0  0.0  15.0   2.25   
32590     13.461538  2.599837  1.0  0.0  0.0  0.0  1.0  0.0   8.0   0.64   
32599     22.596154  3.117780  0.0  0.0  0.0  0.0  1.0  0.0  15.0   2.25   
32603     16.826923  2.822980  0.0  0.0  0.0  0.0  1.0  0.0  11.0   1.21   
32631     32.967033  3.495508  0.0  0.0  0.0  0.0  1.0  0.0  10.0   1.00   

            exp3      exp4    occ occ2    ind ind2  education  
rownames                                                       
500       50.653  187.4161   6800   19    490    2        1.0  
540       17.576   45.6976  1e+05   21  1e+05    5        1.0  
691       24.389   70.7281   4220   14   7070   13        1.0  
843       21.952   61.4656   4020   13   8680   20        1.0  
1775       0.512    0.4096   4230   14   7690   16        1.0  
...          ...       ...    ...  ...    ...  ...        ...  
32580      3.375    5.0625   2010    6   9370   22        1.0  
32590      0.512    0.4096   4720   16   8590   19        1.0  
32599      3.375    5.0625   9620   22   5390    9        1.0  
32603      1.331    1.4641   7150   20   8770   21        1.0  
32631      1.000    1.0000   2920    9   6570   11        1.0  

[1376 rows x 17 columns]

As you can see, education variable now represents the integration of both variables: shs =1 y hsg = 1 . If education = 1, then it means that this person didn't go to college.

In [14]:
Y = np.log2(data['wage']) 
n = len(Y)
z = data.loc[:, ~data.columns.isin(['wage', 'lwage','Unnamed: 0'])]
p = z.shape[1]

print("Number of observation:", n, '\n')
print( "Number of raw regressors:", p)

Number of observation: 1376 

Number of raw regressors: 15


In [15]:
Z_subset = data.loc[:, data.columns.isin(["lwage","sex","mw","so","we","ne","exp1","education"])]
table = Z_subset.mean(axis=0)
table

lwage         2.718562
sex           0.321948
mw            0.286337
so            0.291424
we            0.198401
ne            0.223837
exp1         17.190044
education     1.000000
dtype: float64

In [16]:
table = pd.DataFrame(data=table, columns={"Sample mean":"0"} )
table.index
index1 = list(table.index)
index2 = ["Log Wage","Sex",\
          "Midwest",\
          "South","West","Northeast","Experience","Education"]

In [17]:
table = table.rename(index=dict(zip(index1,index2)))
table

Sample mean
Log Wage       2.718562
Sex            0.321948
Midwest        0.286337
South          0.291424
West           0.198401
Northeast      0.223837
Experience    17.190044
Education      1.000000

In [18]:
print(table.to_latex())

\begin{tabular}{lr}
\toprule
{} &  Sample mean \\
\midrule
Log Wage   &     2.718562 \\
Sex        &     0.321948 \\
Midwest    &     0.286337 \\
South      &     0.291424 \\
West       &     0.198401 \\
Northeast  &     0.223837 \\
Experience &    17.190044 \\
Education  &     1.000000 \\
\bottomrule
\end{tabular}



## Prediction Question

In [19]:
# Import packages for OLS regression
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [20]:
# Models required
basic = 'lwage ~ sex + exp1 + education + mw + so + we + occ2+ ind2'
Y = 'lwage ~  exp1 + education + mw + so + we + occ2+ ind2'
D = 'sex   ~  exp1 + education + mw + so + we + occ2+ ind2'
basic_results_0=smf.ols(Y , data=data)
basic_results = smf.ols(basic , data=data).fit()
print(basic_results.summary()) # estimated coefficients
print( "Number of regressors in the basic model:",len(basic_results.params), '\n')  # number of regressors in the Basic Model

                            OLS Regression Results                            
Dep. Variable:                  lwage   R-squared:                       0.178
Model:                            OLS   Adj. R-squared:                  0.150
Method:                 Least Squares   F-statistic:                     6.268
Date:                Fri, 17 Sep 2021   Prob (F-statistic):           1.38e-32
Time:                        15:02:13   Log-Likelihood:                -874.52
No. Observations:                1376   AIC:                             1843.
Df Residuals:                    1329   BIC:                             2089.
Df Model:                          46                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.5337      0.046     33.459      0.0

##### Note that the basic model consists of $48$ regressors.

In [21]:
# 2. flexible model
flex = 'lwage ~ (exp1 + exp2 + exp3 + exp4 + education + occ2 + ind2 + mw + so + we)**2'
W = 'lwage ~ (exp1 + exp2 + exp3 + exp4 + education + occ2 + ind2 + mw + so + we)**2'
S = 'sex ~ (exp1 + exp2 + exp3 + exp4 + education + occ2 + ind2 + mw + so + we)**2'
flex_results_0 = smf.ols(W , data=data)
flex_results = smf.ols(flex , data=data).fit()
print(flex_results.summary()) # estimated coefficients
print( "Number of regressors in the basic model:",len(flex_results.params), '\n') # number of regressors in the Flexible Model

                            OLS Regression Results                            
Dep. Variable:                  lwage   R-squared:                       0.468
Model:                            OLS   Adj. R-squared:                  0.201
Method:                 Least Squares   F-statistic:                     1.753
Date:                Fri, 17 Sep 2021   Prob (F-statistic):           5.24e-13
Time:                        15:02:14   Log-Likelihood:                -574.91
No. Observations:                1376   AIC:                             2072.
Df Residuals:                     915   BIC:                             4481.
Df Model:                         460                                         
Covariance Type:            nonrobust                                         
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
Intercept                 1.26

Note that the flexible model consists of $826$ regressors.

## Try Lasso next and Partialling out

In [22]:
# Import relevant packages for lasso 
from sklearn.linear_model import LassoCV
from sklearn import linear_model
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error

In [23]:
# Get exogenous variables from flexible model
X = basic_results_0.exog
X.shape

# Set endogenous variable
lwage = data["lwage"]
lwage.shape

(1376,)

In [24]:
sex=data["sex"]
sex

rownames
500      0.0
540      0.0
691      0.0
843      0.0
1775     1.0
        ... 
32580    0.0
32590    1.0
32599    0.0
32603    0.0
32631    0.0
Name: sex, Length: 1376, dtype: float64

In [25]:
alpha = 0.5

In [26]:
# Set penalty value = 0.5
#Lasso Regression
lasso = linear_model.Lasso(alpha = alpha)

#Lasso fitted and predict on Lwage ~ Y
lwage_lasso_fitted = lasso.fit(X, data[['lwage']]).predict(X)

#Get residual from previous model
res= data[['lwage']] - lwage_lasso_fitted.reshape(lwage_lasso_fitted.size ,1)

#Lasso fitted and predict on sex ~ D

basic_results_D = smf.ols(D, data=data)
Sex_lasso = lasso.fit( basic_results_D.exog, data[['sex']] ).predict(basic_results_D.exog)

#Get residual from previous model

res1 = data[['sex']] - Sex_lasso.reshape( Sex_lasso.size, 1)
res2 = pd.DataFrame( np.hstack(( res , res1 )) , columns = ['Res','Res1'] )

#Partialling out

lasso_fit_res = smf.ols( formula = 'Res ~ Res1' , data = res2 ).fit()
coef = lasso_fit_res.summary2().tables[1]['Coef.']['Res1']

# coefficient
print('Lasso Regression: Beta of sex coefficient', coef)

Lasso Regression: Beta of sex coefficient -0.12148020376235977


## Try Lasso next

In [27]:
# Get exogenous variables from flexible model
W = flex_results_0.exog
W.shape

# Set endogenous variable
lwage = data["lwage"]
lwage.shape

(1376,)

In [28]:
sex=data["sex"]
sex.shape

(1376,)

In [29]:
alpha = 0.5

In [30]:
# Set penalty value = 0.5
#Lasso Regression
lasso_1 = linear_model.Lasso(alpha = alpha)

#Lasso fitted and predict on Lwage ~ W
lwage_lasso_fit = lasso_1.fit(X, data[['lwage']]).predict(X)

#Get residual from previous model
res5= data[['lwage']] - lwage_lasso_fit.reshape(lwage_lasso_fit.size ,1)

#Lasso fitted and predict on sex ~ S

basic_results_S = smf.ols(S, data=data)
Sex_lasso_1 = lasso.fit( basic_results_S.exog, data[['sex']] ).predict(basic_results_S.exog)

#Get residual from previous model

res6 = data[['sex']] - Sex_lasso_1.reshape(Sex_lasso_1.size, 1)
res7 = pd.DataFrame( np.hstack((res5,res6)) , columns = ['Res5','Res6'] )

#Partialling out

lasso_fit_res_1 = smf.ols(formula = 'Res5 ~ Res6',data = res7).fit()
coef_1 = lasso_fit_res_1.summary2().tables[1]['Coef.']['Res6']

# coefficient
print('Lasso Regression: Beta of sex coefficient', coef_1)

Lasso Regression: Beta of sex coefficient -0.11870459823135386


###### In both cases, the regression seem pretty similar:

###### Lasso Regression:sex coefficient in Case 1 -0.12148020376235977

###### Lasso Regression: sex coefficient in Case 2 -0.11870459823135386

# Explanation of Sample Splitting

In statistics, you never know what the population looks like, you only have one sample - representative or not - and you can only model according to that sample you have. Splitting is a tool used in data analysis and machine learning, the objective of which is to generalize very well to the population through a model. It consists of randomly separating the sample into two subsets called: train set and test set. It is common to include in the whole train set 75-80% of the total data and on the other hand, 20-25% of towards the test subsample. The importance of this procedure is that it avoids two situations: Overfitting and Underfitting. Overfitting occurs when our machine learns from particular cases and if these do not conform to it, then it will not be able to recognize the new data. On the other hand, underfitting occurs when our model cannot capture the trend, it may be due to the presence of outliers. The steps to follow is to adjust a set of trains, then this model regresses with this new data (test set). And since it's new data, you're basically seeing how your model performs on it.